File: src/main/java/org/khanacademy/logexport/LogsExtractor.java
package org.khanacademy.logexport;

import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.services.bigquery.model.TableFieldSchema;
import com.google.api.services.bigquery.model.TableRow;
import com.google.api.services.bigquery.model.TableSchema;
import com.google.api.services.logging.v2.model.LogEntry;
import com.google.api.services.logging.v2.model.LogLine;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * Top-level logic class for log transformation. The work is subdivided into four top-level
 * components:
 * - The "standard log fields" that are provided directly by GAE, except the app logs.
 * - The app logs.
 * - The event log information, which is extracted from the app logs.
 * - The bingo events, which are also extracted from the app logs.
 */
public class LogsExtractor {
    private final StandardLogFieldParser mStandardLogFieldParser;
    private final AppLogParser mAppLogParser;
    private final EventLogParser mEventLogParser;
    private final BingoEventParser mBingoEventParser;

    public LogsExtractor(StandardLogFieldParser standardLogFieldParser, AppLogParser appLogParser,
                         EventLogParser eventLogParser, BingoEventParser bingoEventParser) {
        mStandardLogFieldParser = standardLogFieldParser;
        mAppLogParser = appLogParser;
        mEventLogParser = eventLogParser;
        mBingoEventParser = bingoEventParser;
    }

    public static LogsExtractor create() {
        return new LogsExtractor(new StandardLogFieldParser(), new AppLogParser(),
                new EventLogParser(), new BingoEventParser());
    }

    public List<TableFieldSchema> getSchemaFields() {
        List<TableFieldSchema> fields = new ArrayList<>();
        fields.addAll(mStandardLogFieldParser.getSchemaFields());
        fields.add(mAppLogParser.getSchemaField());
        fields.addAll(mEventLogParser.getSchemaFields());
        fields.addAll(mBingoEventParser.getSchemaFields());
        return fields;
    }

    public TableSchema getSchema() {
        return new TableSchema().setFields(getSchemaFields());
    }

    public TableRow extractLogs(LogEntry logEntry) throws IOException {
        TableRow row = new TableRow();
        mStandardLogFieldParser.populateStandardLogFields(row, logEntry);
        List<LogLine> logLines = mAppLogParser.getLogLines(logEntry);
        mAppLogParser.populateAppLogField(row, logLines);
        mEventLogParser.populateEventLogFields(row, logLines);
        mBingoEventParser.populateBingoEventFields(row, logLines);
        return row;
    }
}

File: src/main/java/org/khanacademy/logexport/LogsExtractor.kt
package org.khanacademy.logexport

import com.google.api.services.bigquery.model.TableFieldSchema
import com.google.api.services.bigquery.model.TableRow
import com.google.api.services.bigquery.model.TableSchema
import com.google.api.services.logging.v2beta1.model.LogEntry

import java.io.IOException
import java.util.ArrayList

/**
 * Top-level logic class for log transformation. The work is subdivided into four top-level
 * components:
 * - The "standard log fields" that are provided directly by GAE, except the app logs.
 * - The app logs.
 * - The event log information, which is extracted from the app logs.
 * - The bingo events, which are also extracted from the app logs.
 */
class LogsExtractor(private val mStandardLogFieldParser: StandardLogFieldParser, private val mAppLogParser: AppLogParser,
                    private val mEventLogParser: EventLogParser, private val mBingoEventParser: BingoEventParser) {

    val schema: TableSchema
        get() {
            val fields = ArrayList<TableFieldSchema>()
            fields.addAll(mStandardLogFieldParser.schemaFields)
            fields.add(mAppLogParser.schemaField)
            fields.addAll(mEventLogParser.schemaFields)
            fields.addAll(mBingoEventParser.schemaFields)
            return TableSchema().setFields(fields)
        }

    @Throws(IOException::class)
    fun extractLogs(logEntry: LogEntry): TableRow {
        val row = TableRow()
        mStandardLogFieldParser.populateStandardLogFields(row, logEntry)
        val logLines = mAppLogParser.getLogLines(logEntry)
        mAppLogParser.populateAppLogField(row, logLines)
        mEventLogParser.populateEventLogFields(row, logLines)
        mBingoEventParser.populateBingoEventFields(row, logLines)
        return row
    }

    companion object {

        fun create(): LogsExtractor {
            return LogsExtractor(StandardLogFieldParser(), AppLogParser(),
                    EventLogParser(), BingoEventParser())
        }
    }
}
